README.md
# basicModal

Easy-to-use dialog system for modern web-apps.

![Modal Screenshot](https://l.electerious.com/uploads/big/0aac1124d3884e2ac0d62f0879d27333.png)

basicModal is a dialog-system for modern web-apps. It includes everything you need to display information, ask questions or request input from the user. Dialogs can be chained, so you can easily build a setup-assistant or show messages in a predefined order. Invalid input can be highlighted and handled using the included function.

basicModal is written in Vanilla JS and has zero dependencies. It uses [SASS](http://sass-lang.com) and [Flexbox](http://dev.w3.org/csswg/css-flexbox/).

Tested with the latest versions of [Mozilla Firefox](https://www.mozilla.org/en-US/firefox/new/), [Apple Safari](https://www.apple.com/safari/), [Google Chrome](https://www.google.com/chrome/browser/), [Microsoft Internet Explorer](http://windows.microsoft.com/en-us/internet-explorer/download-ie) (10+) and [Opera](https://www.opera.com/).

## Contents

- [Demos](#demos)
- [Features](#features)
- [Requirements](#requirements)
- [Setup](#setup)
- [Usage](#usage)
	- [Alert](#alert)
	- [Prompt](#prompt)
	- [Input](#input)
- [Functions](#functions)
	- [Show](#show)
	- [Error](#error)
	- [Visible](#visible)
	- [Action & Cancel](#action--cancel)
	- [Reset](#reset)
	- [Get Values](#get-values)
	- [Close](#close)
- [Options](#options)

## Demos

| Name | Description | Link |
|:-----------|:------------|:------------|
| Alert | Modal similar to `alert()`. The modal contains a message and a button. The message can be filled with HTML and the button fires the specified function when pressed. | [Demo](https://basicmodal.electerious.com/#alert) |
| Prompt | The prompt dialog is great when you want a decision or answer from the user. | [Demo](https://basicmodal.electerious.com/#prompt) |
| Login | Building a login with basicModal is super easy. It includes everything you need, like the ability to highlight invalid input. | [Demo](https://basicmodal.electerious.com/#login) |

## Features

- Works in all modern browsers
- Zero dependencies
- CommonJS and AMD support
- Support for text inputs
- Highlight invalid input
- Execute dialogs in row

## Requirements

basicModal depends on the following browser APIs:

- [classList](http://caniuse.com/#feat=classlist)
- [Flexible Box Layout Module](http://caniuse.com/#feat=flexbox)

Some of these APIs are capable of being polyfilled in older browsers. Check the linked resources above to determine if you must polyfill to achieve your desired level of browser support.

## Setup

We recommend to install basicModal using [Bower](https://bower.io/) or [npm](https://npmjs.com).

```sh
bower install basicModal
```
```sh
npm install basicmodal
```

Include the CSS file in the `head` tag and the JS file at the end of your `body` tag…

```html
<link rel="stylesheet" href="dist/basicModal.min.css">
```
```html
<script src="dist/basicModal.min.js"></script>
```

…or skip the JS file and use basicModal as a module:

```js
const basicModal = require('basicmodal')
```

## Usage

### Alert

Lets start with a modal similar to `alert()`. The modal contains a message and a button. The message can be filled with HTML and the button fires the specified function when pressed.

```js
basicModal.show({
	body: '<p>This is a dead-simple alert modal!<br>The message can be filled with anything you want.</p>',
	buttons: {
		action: {
			title: 'Dismiss',
			fn: basicModal.close
		}
	}
})
```

### Prompt

The prompt dialog is great when you want a decision or answer from the user. The only difference from the first example is the additional button.

```js
basicModal.show({
	body: '<p>This type of modal can be used to ask the user questions. Are you sure you want to continue?</p>',
	buttons: {
		cancel: {
			title: 'Cancel',
			fn: basicModal.close
		},
		action: {
			title: 'Continue',
			fn: basicModal.close
		}
	}
})
```

### Input

Building an input-dialog with basicModal is super easy. It includes everything you need, like the ability to highlight invalid fields. The specified action-button-function receives an object which includes the values of all inputs. Use the name attribute in your HTML to set the name of the inputs.

```js
basicModal.show({
	body: '<p>This type of modal can be used to ask the user questions. Please enter your name:</p><input class="basicModal__text" type="text" name="name" placeholder="Jane Doe">',
	buttons: {
		cancel: {
			title: 'Cancel',
			fn: basicModal.close
		},
		action: {
			title: 'Continue',
			fn: (data) => {

				if (data.name.length===0) return basicModal.error('name')

				console.log(data)
				basicModal.close()

			}
		}
	}
})
```

## Functions

basicModal comes with a handful of handy functions. Below are all of them along with a short description.

### Show

The most important function of basicModal. Call `show` to show a modal. The `object` you pass to the function includes all the information about the modal. Take a look at the demos above to get a feel of the capabilities.

```ts
basicModal.show(options: object)
```

### Error

Use the `error` function to highlight invalid input. This function can only be used when the modal has input elements. Each input needs a `name` attribute as shown in the example above. The input with the matching attribute will be marked red and the modal will shake to signal an error. This function also executes the `reset` function, to remove previous errors and to reactive the buttons.

```ts
basicModal.error(nameAttribute: string)
```

### Visible

Check if there's a visible modal. Returns `true` when a modal is visible and `false` otherwise.

```ts
basicModal.visible() : boolean
```

### Action & Cancel

You can trigger the buttons of the modal manually if wanted. Triggering a button is equivalent to clicking them.

```ts
basicModal.action() : boolean
```

```ts
basicModal.cancel() : boolean
```

### Reset

In order to avoid multiple executions of the same function, buttons can only be pressed once before being disabled. Use reset if you want to reactivate the buttons or to reset the highlighted input errors.

```ts
basicModal.reset() : boolean
```

### Get Values

The following function returns an object, which includes all input values from the current modal.

```ts
basicModal.getValues() : object
```

### Close

And finally: Close the modal. You can force close a modal by passing true to the function. This is required when the modal is created with the closable property set to `false`.

```ts
basicModal.close(forceClose: boolean) : boolean
```

## Options

List of options you can pass to the `basicModal.show` function:

```js
basicModal.show({

	// String containing HTML (required)
	body: '<p>String containing HTML</p>',

	// String - List of custom classes added to the modal (optional)
	class: 'customClass01 customClass02',

	// Boolean - Define if the modal can be closed with the close-function (optional)
	closable: true,

	// Function - Will fire when modal is visible (optional)
	callback: undefined,

	// Object - basicModal accepts up to two buttons and requires at least one of them
	buttons: {

		cancel: {

			// String (optional)
			title: 'Cancel',

			// String - List of custom classes added to the button (optional)
			class: 'customButtonClass',

			// Function - Will fire when user clicks the button (required)
			fn: basicModal.close

		},

		action: {

			// String (optional)
			title: 'OK',

			// String - List of custom classes added to the button (optional)
			class: 'customButtonClass',

			// Function - Will fire when user clicks the button (required)
			fn: (data) => basicModal.close()

		}

	}

})
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.3.9] - 2017-08-08

### New

- Added a changelog

### Changed

- Ignore `yarn.lock` and `package-lock.json` files

src/scripts/basicModal.js
let lastFocus = null

const THEME = {
	small  : 'basicModal__small',
	xclose : 'basicModal__xclose'
}

const dom = function(elem = '', multiple = false) {

	if (multiple===true) return document.querySelectorAll('.basicModal ' + elem)
	else                 return document.querySelector('.basicModal ' + elem)

}

const each = (data, fn) => {

	if ((data).constructor===Object) return [].forEach.call(Object.keys(data), (key) => fn(data[key], key, data))
	else                             return [].forEach.call(data, (item, i) => fn(item, i, data))

}

const valid = function(data) {

	if (data==null || Object.keys(data).length===0) {
		console.error('Missing or empty modal configuration object')
		return false
	}

	if (data.body==null)       data.body = ''
	if (data.class==null)      data.class = ''
	if (data.closable!==false) data.closable = true

	if (data.buttons==null) {
		console.error('basicModal requires at least one button')
		return false
	}

	// Validate action-button
	if (data.buttons.action!=null) {

		if (data.buttons.action.class==null) data.buttons.action.class = ''
		if (data.buttons.action.title==null) data.buttons.action.title = 'OK'

		if (data.buttons.action.fn==null) {
			console.error('Missing fn for action-button')
			return false
		}

	}

	// Validate cancel-button
	if (data.buttons.cancel!=null) {

		if (data.buttons.cancel.class==null) data.buttons.cancel.class = ''
		if (data.buttons.cancel.title==null) data.buttons.cancel.title = 'Cancel'

		if (data.buttons.cancel.fn==null) {
			console.error('Missing fn for cancel-button')
			return false
		}

	}

	return true

}

const build = function(data) {

	let icon = '<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="0 0 512 512"><path d="M405 136.798l-29.798-29.798-119.202 119.202-119.202-119.202-29.798 29.798 119.202 119.202-119.202 119.202 29.798 29.798 119.202-119.202 119.202 119.202 29.798-29.798-119.202-119.202z"/></svg>',
	    html = ''

	html += `
	        <div class='basicModalContainer basicModalContainer--fadeIn' data-closable='${ data.closable }'>
	            <div class='basicModal basicModal--fadeIn ${ data.class }' role="dialog">
	                <div class='basicModal__content'>
	                    ${ data.body }
	                </div>
	                <div class='basicModal__buttons'>
	        `

	// Cancel-button
	if (data.buttons.cancel!=null) {
		if (data.buttons.cancel.class.indexOf('basicModal__xclose')===-1) {

			// Default close-button
			html +=	`<a id='basicModal__cancel' class='basicModal__button ${ data.buttons.cancel.class }'>${ data.buttons.cancel.title }</a>`

		} else {

			// Custom close-button for the login-theme
			html += `<div id='basicModal__cancel' class='basicModal__button ${ data.buttons.cancel.class }' aria-label='close'>${ icon }</div>`

		}
	}

	// Action-button
	if (data.buttons.action!=null) {

		html += `<a id='basicModal__action' class='basicModal__button ${ data.buttons.action.class }'>${ data.buttons.action.title }</a>`

	}

	html += `
	                </div>
	            </div>
	        </div>
	        `

	return html

}

const getValues = function() {

	let values = null,
	    inputs = dom('input', true)

	if (inputs.length>0) {

		values = {}

		// Get value from all inputs
		each(inputs, (input) => {

			let name  = input.getAttribute('name'),
			    value = input.value

			// Store name and value of input
			if (name!=null) values[name] = value

		})

		// Set value back to null when object empty
		if (Object.keys(values).length===0) values = null

	}

	return values

}

const bind = function(data) {

	// Cancel-button
	if (data.buttons.cancel!=null) {

		dom('#basicModal__cancel').onclick = function() {

			// Don't execute function when button has been clicked already
			if (this.classList.contains('basicModal__button--active')===true) return false

			this.classList.add('basicModal__button--active')
			data.buttons.cancel.fn()

		}

	}

	// Action-button
	if (data.buttons.action!=null) {

		dom('#basicModal__action').onclick = function() {

			// Don't execute function when button has been clicked already
			if (this.classList.contains('basicModal__button--active')===true) return false

			this.classList.add('basicModal__button--active')
			data.buttons.action.fn(getValues())

		}

	}

	// Bind inputs
	each(dom('input', true), (input) => {

		input.onkeydown = input.onblur = function() { this.classList.remove('error') }

	})

	return true

}

const show = function(data) {

	// Validate data and set default values
	if (valid(data)===false) return false

	// Close open modal
	if (dom()!=null) {
		close(true)
		setTimeout(() => show(data), 301)
		return false
	}

	// Save focused element
	lastFocus = document.activeElement

	// Build modal
	let html = build(data)

	// Append modal to DOM
	document.body.insertAdjacentHTML('beforeend', html)

	// Bind elements
	bind(data)

	// Select the first input when available
	let input = dom('input')
	if (input!=null) input.select()

	// Execute callback when available
	if (data.callback!=null) data.callback(data)

	return true

}

const error = function(input) {

	// Reactive buttons and remove old errors
	reset()

	let elem = dom(`input[name='${ input }']`)

	elem.classList.add('error')
	elem.select()

	// Shake input
	dom().classList.remove('basicModal--fadeIn', 'basicModal--shake')
	setTimeout(() => dom().classList.add('basicModal--shake'), 1)

}

const visible = function() {

	if (dom()!=null) return true
	return false

}

const action = function() {

	let elem = dom('#basicModal__action')

	if (elem!=null) {

		elem.click()
		return true

	}

	return false

}

const cancel = function() {

	let elem = dom('#basicModal__cancel')

	if (elem!=null) {

		elem.click()
		return true

	}

	return false

}

const reset = function() {

	// Reactive buttons
	let buttons = dom('.basicModal__button', true)
	each(buttons, (button) => button.classList.remove('basicModal__button--active'))

	// Remove errors
	let inputs = dom('input', true)
	each(inputs, (input) => input.classList.remove('error'))

	return true

}

const close = function(force) {

	// Only close when a modal is visible
	if (visible()===false) return false

	// Validate force
	if (force!==true) force = false

	// Get modal container
	let container = dom().parentElement

	// Don't close when modal not closable
	// Use 'force===true' to close unclosebale modal
	if (container.getAttribute('data-closable')==='false' && force===false) return false

	container.classList.remove('basicModalContainer--fadeIn')
	container.classList.add('basicModalContainer--fadeOut')

	setTimeout(() => {

		// Only close when container exists
		if (container==null)               return false
		if (container.parentElement==null) return false

		container.parentElement.removeChild(container)

	}, 300)

	// Restore last active element
	if (lastFocus!=null) {
		lastFocus.focus()
		lastFocus = null
	}

	return true

}

return {
	THEME,
	show,
	visible,
	getValues,
	action,
	cancel,
	error,
	reset,
	close
}

src/scripts/main.js
let lastFocus = null

export const THEME = {
	small  : 'basicModal__small',
	xclose : 'basicModal__xclose'
}

const dom = function(elem = '', multiple = false) {

	if (multiple===true) return document.querySelectorAll('.basicModal ' + elem)
	else                 return document.querySelector('.basicModal ' + elem)

}

const each = function(data, fn) {

	if (data==null) return false

	if ((data).constructor===Object) return Array.prototype.forEach.call(Object.keys(data), (key) => fn(data[key], key, data))
	else                             return Array.prototype.forEach.call(data, (item, i) => fn(item, i, data))

}

const valid = function(data) {

	if (data==null || Object.keys(data).length===0) {
		console.error('Missing or empty modal configuration object')
		return false
	}

	if (data.body==null)       data.body = ''
	if (data.class==null)      data.class = ''
	if (data.closable!==false) data.closable = true

	if (data.buttons==null) {
		console.error('basicModal requires at least one button')
		return false
	}

	// Validate action-button
	if (data.buttons.action!=null) {

		if (data.buttons.action.class==null) data.buttons.action.class = ''
		if (data.buttons.action.title==null) data.buttons.action.title = 'OK'

		if (data.buttons.action.fn==null) {
			console.error('Missing fn for action-button')
			return false
		}

	}

	// Validate cancel-button
	if (data.buttons.cancel!=null) {

		if (data.buttons.cancel.class==null) data.buttons.cancel.class = ''
		if (data.buttons.cancel.title==null) data.buttons.cancel.title = 'Cancel'

		if (data.buttons.cancel.fn==null) {
			console.error('Missing fn for cancel-button')
			return false
		}

	}

	return true

}

const build = function(data) {

	let icon = '<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="0 0 512 512"><path d="M405 136.798l-29.798-29.798-119.202 119.202-119.202-119.202-29.798 29.798 119.202 119.202-119.202 119.202 29.798 29.798 119.202-119.202 119.202 119.202 29.798-29.798-119.202-119.202z"/></svg>'
	let html = ''

	html += `
	        <div class='basicModalContainer basicModalContainer--fadeIn' data-closable='${ data.closable }'>
	            <div class='basicModal basicModal--fadeIn ${ data.class }' role="dialog">
	                <div class='basicModal__content'>
	                    ${ data.body }
	                </div>
	                <div class='basicModal__buttons'>
	        `

	// Cancel-button
	if (data.buttons.cancel!=null) {
		if (data.buttons.cancel.class.indexOf('basicModal__xclose')===-1) {

			// Default close-button
			html +=	`<a id='basicModal__cancel' class='basicModal__button ${ data.buttons.cancel.class }'>${ data.buttons.cancel.title }</a>`

		} else {

			// Custom close-button for the login-theme
			html += `<div id='basicModal__cancel' class='basicModal__button ${ data.buttons.cancel.class }' aria-label='close'>${ icon }</div>`

		}
	}

	// Action-button
	if (data.buttons.action!=null) {

		html += `<a id='basicModal__action' class='basicModal__button ${ data.buttons.action.class }'>${ data.buttons.action.title }</a>`

	}

	html += `
	                </div>
	            </div>
	        </div>
	        `

	return html

}

export const getValues = function() {

	let values  = {}
	let inputs  = dom('input[name]', true)
	let selects = dom('select[name]', true)

	// Get value from all inputs
	each(inputs, (input) => {

		let name  = input.getAttribute('name')
		let value = input.value

		// Store name and value of input
		values[name] = value

	})

	// Get selected value from all selects
	each(selects, (select) => {

		let name  = select.getAttribute('name')
		let value = select.options[select.selectedIndex].value

		// Store name and value of select
		values[name] = value

	})

	return (Object.keys(values).length===0 ? null : values)

}

const bind = function(data) {

	// Cancel-button
	if (data.buttons.cancel!=null) {

		dom('#basicModal__cancel').onclick = function() {

			// Don't execute function when button has been clicked already
			if (this.classList.contains('basicModal__button--active')===true) return false

			this.classList.add('basicModal__button--active')
			data.buttons.cancel.fn()

		}

	}

	// Action-button
	if (data.buttons.action!=null) {

		dom('#basicModal__action').onclick = function() {

			// Don't execute function when button has been clicked already
			if (this.classList.contains('basicModal__button--active')===true) return false

			this.classList.add('basicModal__button--active')
			data.buttons.action.fn(getValues())

		}

	}

	// Bind inputs
	each(dom('input', true), (input) => {

		input.oninput = input.onblur = function() { this.classList.remove('error') }

	})

	// Bind selects
	each(dom('select', true), (select) => {

		select.onchange = select.onblur = function() { this.classList.remove('error') }

	})

	return true

}

export const show = function(data) {

	// Validate data and set default values
	if (valid(data)===false) return false

	// Close open modal
	if (dom()!=null) {
		close(true)
		setTimeout(() => show(data), 301)
		return false
	}

	// Save focused element
	lastFocus = document.activeElement

	// Build modal
	let html = build(data)

	// Append modal to DOM
	document.body.insertAdjacentHTML('beforeend', html)

	// Bind elements
	bind(data)

	// Select the first input when available
	let input = dom('input')
	if (input!=null) input.select()

	// If there is no input but a select, select it
	let select = dom('select')
	if (input==null && select!=null) select.focus()

	// Execute callback when available
	if (data.callback!=null) data.callback(data)

	return true

}

export const error = function(nameAttribute) {

	// Reactive buttons and remove old errors
	reset()

	// Select element with the given name attribute
	let elem = dom(`input[name='${ nameAttribute }']`) || dom(`select[name='${ nameAttribute }']`)

	// Stop function when element not found
	if (elem==null) return false

	elem.classList.add('error')

	if (typeof elem.select==='function') elem.select()
	else                                 elem.focus()

	// Shake input or select
	dom().classList.remove('basicModal--fadeIn', 'basicModal--shake')
	setTimeout(() => dom().classList.add('basicModal--shake'), 1)

}

export const visible = function() {

	if (dom()!=null) return true
	return false

}

export const action = function() {

	let elem = dom('#basicModal__action')

	if (elem!=null) {

		elem.click()
		return true

	}

	return false

}

export const cancel = function() {

	let elem = dom('#basicModal__cancel')

	if (elem!=null) {

		elem.click()
		return true

	}

	return false

}

export const reset = function() {

	// Reactive buttons
	let buttons = dom('.basicModal__button', true)
	each(buttons, (button) => button.classList.remove('basicModal__button--active'))

	// Remove errors from inputs
	let inputs = dom('input', true)
	each(inputs, (input) => input.classList.remove('error'))

	// Remove errors from selects
	let selects = dom('select', true)
	each(selects, (select) => select.classList.remove('error'))

	return true

}

export const close = function(force = false) {

	// Only close when a modal is visible
	if (visible()===false) return false

	// Get modal container
	let container = dom().parentElement

	// Don't close when modal not closable
	// Use 'force===true' to close unclosebale modal
	if (container.getAttribute('data-closable')==='false' && force===false) return false

	container.classList.remove('basicModalContainer--fadeIn')
	container.classList.add('basicModalContainer--fadeOut')

	setTimeout(() => {

		// Only close when container exists
		if (container==null)               return false
		if (container.parentElement==null) return false

		container.parentElement.removeChild(container)

	}, 300)

	// Restore last active element
	if (lastFocus!=null) {
		lastFocus.focus()
		lastFocus = null
	}

	return true

}